Windows Support for CLI interface

kangcliff opened the pull request:
Implements the underlying device IO interface for Windows in _hidapi_ submodule, as I wasn't able to get the devices working with [#1971](https://github.com/pwr-Solaar/Solaar/pull/1971). For now only the CLI interface is working as getting GTK to work on Windows requires the mingw environment. Devices I tested:

- MX Master 3 for Business
- MX Keys for Business

 (with both Bolt receiver and Bluetooth)
_Note: Requires building cython-hidapi dependency from source as I also pushed there for the missing bus_type field, changes not on pip yet._

kangcliff committed:
Initial Windows support.

kangcliff committed:
Initial Windows support.

kangcliff committed:
Initial Windows support.

kangcliff committed:
Initial Windows support.

kangcliff committed:
Restoring commented out imports.

kangcliff committed:
Restoring commented out imports.

kangcliff committed:
Restoring commented out imports.

kangcliff committed:
Fixing module related errors.

kangcliff committed:
Fixing module related errors.

kangcliff committed:
Fixing module related errors.

kangcliff committed:
Fixed a bug when reading from device. Updating setup script.

kangcliff committed:
Fixed a bug when reading from device. Updating setup script.

kangcliff committed:
Fixed a bug when reading from device. Updating setup script.

pfps commented:
This is going to require some investigation.

pfps commented:
Have to talked to @markopy about hidapi?

pfps commented:
The style checks appear to have failed.  Can you look into this?   Ideally you should run the style checks locally as part of your local commit process.

pfps committed:
release 1.1.11rc1

kangcliff committed:
Fix for styling mistakes.

kangcliff committed:
Fix for styling mistakes.

kangcliff committed:
Fix for styling mistakes.

kangcliff committed:
Merge branch 'master' into win_support

markopy committed:
Partial support for macOS and Windows (#1971)

* Add support for macOS via hidapi

* Style fixes

* Ignore keyboard and mouse input devices

* Don't require pyudev on mac and windows

* Fix debug log format error

* More logging for failed hidpp checks

* Don't try to load hid_darwin_set_open_exclusive on windows

* Bring back button for rule editor since some rules will work

---------

Co-authored-by: markopy <(none)>
Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

pfps committed:
rules: style fix

pfps commented:
@kangcliff PR #1971 was merged.  Can you update this PR to work off the current master?

kangcliff committed:
Merge branch 'win_support' of https://github.com/kangcliff/Solaar into win_support

kangcliff commented:
@pfps I rebased it on current master, thinking now with the mac support MR being merged, it seems like a better idea to rework that to be based on cython_hidapi? That would be easier to read through than relying on ctypes loading the hidapi C library.

pfps commented:
It looks as if a style check failed.

Can you also try to get the MacOS commits removed so that it is easy to see the changes you made?

Your change sounds reasonable.   Give it a try.

Commit:
device: add K845 keyboard

pfps committed:
device: remove incorrect feature for M325 mice

pfps committed:
docs: add files for MX Anywhere 2S and G915

pfps committed:
gui: defend against lightspeed receivers that contact devices for basic information

pfps committed:
docs: add file for G915

txelu committed:
po: Spanish translation reviewed (#2209)

pfps committed:
device: add names for new Logitech features

pfps committed:
udev: report hidraw node in debugging messages

pfps committed:
docs: add descriptions of M650 and PRO X 2

pfps committed:
cli: be defensive in device error message

pfps committed:
rules: write empty file if there are no rules to save

pfps committed:
device: add extended report rate setting

pfps committed:
device: support bug in backlight on MX Keys S

pfps committed:
device: support backlight levels and duration

pfps committed:
ui: use Report Rate instead of Polling for movement report rate

pfps committed:
docs: add descriptions of G305 and MX Keys S

pfps committed:
ui: downgrade assertion on missing notification flag to warning

pfps committed:
devices: override name of Candy Companion Chip

antonsoroko committed:
docs: Update installation.md with new udev rules location (#2234)

location was changed in 
https://github.com/pwr-Solaar/Solaar/commit/cf71736920ff85e9f5a082d3a7cfaadfc00af13d

antonsoroko committed:
release: fix changelog parsing in release.sh after d57af51316799252fdc67f575dc72d2568b3e48c (#2235)

antonsoroko committed:
release: Add stable branch to release.sh (#2236)

* Add stable branch to release.sh

To be used in PPA builds.

* do not update stable branch if prerelease

---------

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

pfps committed:
cli: no numbers for USB and Bluetooth devices

pfps committed:
device: support onboard profiles

antonsoroko committed:
po: Mention language pack for Gnome in i18n.md (#2241)

* mention language pack for Gnome in i18n.md

* i18n.md: Poedit and Lokalize

antonsoroko committed:
po: Update RU language translation (#2242)

* refresh ru.po pointers

* format po/ru.po to see changes w/o changed spaces

* Update translation in po/ru.po

pfps committed:
ui: upgrade onboard profiles setting to allow profile selection

pfps committed:
ui: handle onboard profiles notifications

pfps committed:
ui: handle backlight notification

pfps committed:
ui: add setting to change LED control between firmware and software

pfps committed:
cli: report more information on exception when loading profiles

pfps committed:
device: decipher LED control info in profiles

pfps committed:
device: read profiles from ROM if none in Flash

pfps committed:
device: remove extra debugging for backlight

pfps committed:
device: add version and device name to profiles

pfps committed:
device: better handing of unknown values in profiles

pfps committed:
device: add structures for LED control

kocoman2 commented:
how to use cli version for windows? thx

pfps commented:
> how to use cli version for windows? thx

@kangcliff Can you answer this?

pfps committed:
device: add settings for LED Zone control

pfps committed:
ui: add UI for LED Zone control

pfps committed:
device: handle v4 of profiles data

pfps committed:
docs: document profiles

pfps committed:
ui: better startup behavior for LED effect settings

pfps committed:
device: clean up data for LED effects

pfps committed:
device: remove dependency on webcolors

pfps committed:
release 1.1.11rc1

Matthaiks committed:
po: Update Polish translation (#2252)

pfps committed:
cli: catch assertion errors when reading setting values from devices

MattHag committed:
tests: Introduce tests with pytest

Relates #1097

MattHag committed:
tests: Add GitHub action for tests

Relates #1097

pfps committed:
ui: don't lock setting when an error occurs

pfps committed:
ui: update label and tooltip for divert-gkeys setting

pfps committed:
ui: update label and tooltip for divert-gkeys setting

pfps committed:
ui: refactor code to record change to setting

pfps committed:
ui: don't unlock setting when changed by external means

pfps committed:
device: correctly handle profile button with no action

MattHag committed:
logging: Simplify logger instantiation

* logging: Simplify logger instantiation

Relates #2254

* logging: Remove aliases

Relates #2254

* logging: Replace deprecated warn with warning

Related #2254

* logging: Fix mistake

Related #2257

pfps committed:
release 1.1.11rc2

pfps committed:
device: improve imports in logitech_receiver

device: move some imports to top of modules

device: break up imports loop with device descriptors

device: break up imports loop by moving a function from notifications.py to setting_templates.py

device: break import loop between device.py and diversion.py by using device to access method

pfps committed:
cli: move imports in __init__.py to top of file

pfps committed:
ui: improve imports in ui

ui: move imports in about.py to top of file

ui: move imports to top of notify.py

ui: move imports to top of window.py

ui: reorder imports at beginning of __init__.py

ui: move imports to top of tray.py

ui: move common code out of __init__.py to common.py

ui: move imports to top of __init___.py

pfps committed:
solaar: improve imports and guard Gtk, etc imports with correct version

solaar: move imports to top of files

solaar: move more imports to top of files

solaar: guard Gtk, etc imports with correct version

pfps committed:
device: improve imports in logitech_receiver

device: move imports of ui modules to beginning of files

logitech_receiver: remove imports from __init__.py

pfps committed:
ui: tidy up scrolling appearance in configuration panel

pfps committed:
doc: add document on implementation

MattHag committed:
udev: Simplify installation of udev rules

* Simplify installation of udev rules

Fixes #2263

* udev: Shorten udev installation description

Related #2263

* udev: Shorten udev installation description

Related #2263

* udev: Update installation of udev rules

Related #2263

* Update docs/installation.md

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

* Update Makefile

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

* Update Makefile

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

* Update docs/installation.md

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

* Update docs/installation.md

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

* Update docs/installation.md

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

---------

Co-authored-by: Peter F. Patel-Schneider <pfpschneider@gmail.com>

MattHag committed:
Fix invalid func name set logo name

Related #2254, #2276

pfps committed:
device: improve features array

Matthaiks committed:
po: Update Polish translation (#2275)

pfps committed:
ui: better checking for setting in record_setting

pfps committed:
release 1.1.11rc3

MattHag committed:
Setup macOS tests using GitHub action (#2284)

Related #1244

MattHag committed:
logitech_receiver: Move hidpp10 constants into new module

Related #1097

MattHag committed:
logitech_receiver: Move exceptions into own module

Related #1097

MattHag committed:
logitech_receiver: Move hidpp20 constants into new module

Related #1097

pfps committed:
dist: rename light icons and install them in correct place

MattHag committed:
macos: Remove dbus from macos dependencies

Related #2284

pfps committed:
release 1.1.11rc4

pfps committed:
dist: correctly find light icons

pfps committed:
device: streamline status code

proletarius101 committed:
dist: Add the <developer/> tag in the metainfo



Flathub requires this tag: https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#name-summary-and-developer-name.

The format of this tag is defined in https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-developer.

pfps committed:
release 1.1.11

MattHag committed:
tests: Add tests of common module

Introduces unit tests for Solaar.

Related #1097

MattHag committed:
tests: Show pytest coverage in GitHub CI tests

Related #1097

pfps committed:
ui: move ui_async to common.py

MattHag committed:
dist: extend makefile and tests

* Show pytest coverage in GitHub CI tests

Related #1097

* Extend Makefile with installation and test targets

Refactor setup steps to unify commands between Linux and macOS.
Move bash commands into Makefile for consistency and enable local
execution of GitHub CI commands corresponding Makefile targets.

Install on Ubuntu:
make install_ubuntu

Install on Ubuntu for development:
make install_ubuntu PIP_ARGS=."[test]"

Fixes #2303

* Improve name of GitHub test actions

Related #2303

MattHag committed:
dist: Fix deprecated GitHub actions

* Show pytest coverage in GitHub CI tests

Related #1097

* Extend Makefile with installation and test targets

Refactor setup steps to unify commands between Linux and macOS.
Move bash commands into Makefile for consistency and enable local
execution of GitHub CI commands corresponding Makefile targets.

Install on Ubuntu:
make install_ubuntu

Install on Ubuntu for development:
make install_ubuntu PIP_ARGS=."[test]"

Fixes #2303

* Improve name of GitHub test actions

Related #2303

* Upgrade GitHub actions to Node.js 20

Replaces deprecated Node.js 16 actions.

Related #2256, #2284

pfps committed:
device: add callback to call when changing a setting

pfps committed:
ui: implement setting_changed callback and pass in to new devices and receivers

pfps committed:
hidapi: upgrade debugging in udev

pfps committed:
logitech_receiver: style fixes

pfps commented:
@kangcliff There are some updates that are needed for this to fit with the current version of Solaar.

kangcliff committed:
Fix for conflicts.

kangcliff committed:
Merge branch 'win_support' of https://github.com/kangcliff/Solaar into win_support

pfps commented:
Thanks, except that this PR now has a lot of PRs that have been merged into Solaar.   I'm not sure how this happened.

vulpes2 commented:
Does #1971 supersede this PR?

pfps commented:
@vulpes2 Unknown.  I don't have either a Mac or a Windows installation to check.

ishay3000 commented:
Apologies for bumping this thread. Has there been any progress with the windows port since the last update? Thank you all for your work!

pfps commented:
The work needs someone who uses and understands Windows.  So little is being done.
